### PR TITLE
bugfix: fix `typeDefinition` on backticked identifier

### PR DESCRIPTION
**Pull request by harpocrates**
One of the short-circuiting conditions in the type definition logic was a bit too broad, leading to `typeDefinition` not returning a result if the position queried happened to precede a space in an identifier (see the test that was added).

### CODE REVIEW

**Comment by harpocrates** [comment]
I think failures are unrelated...

**Review by kasiaMarek — COMMENTED** [review]
Can you add a few tests for whitespaces/delimiters? Just to make sure, that none of the fallbacks in the logic will make it point to some random, near by symbol. E.g.:
```
val f = x   @@  + x
```

**Comment by harpocrates** [comment]
@kasiaMarek oh darn those are problematic.

```
val f = x   @@  + x
```

is going to locate `f.+` as the enclosing tree. Would it be better to try to tokenize the file and replace the prior whitespace check directly occurring on the source code with a check about the cursor position being inside a token?

**Comment by kasiaMarek** [comment]
> Would it be better to try to tokenize the file and replace the prior whitespace check directly occurring on the source code with a check about the cursor position being inside a token?

That sounds reasonable. This will probably be slightly slower, since you tokenise the whole file but I don't think there will much of an impact. On the other hand, people will rarely do `go to def` on white spaces, so you can postpone that check and e.g. tokenise only the result of `definitionTypedTreeAt` which will be smaller than the whole file.

**Comment by harpocrates** [comment]
@kasiaMarek Updated - thanks for the pointer. I opted to do the tokenization inside `definitionTypedTreeAt` just because there's logic towards the end of that function (around the import parts handling) that results in the returned `Tree` not always having a position attached to it.

File: mtags/src/main/scala-2/scala/meta/internal/pc/PcDefinitionProvider.scala [hunk 1]
```diff
@@ -185,4 +188,40 @@ class PcDefinitionProvider(val compiler: MetalsGlobal, params: OffsetParams) {
...
+   *
+   * @param offset
+   * @param treeEnclosingOffset
+   * @return
+   */
+  private def tokensAtOffsetSupportDefinition(
+      offset: Int,
+      treeEnclosingOffset: Tree
+  ): Boolean = {
+    val treePos = treeEnclosingOffset.pos
```
Comment (tgodzik) [inline mtags/src/main/scala-2/scala/meta/internal/pc/PcDefinitionProvider.scala:206]: Wouldn't it be simpler to match tree to one of `Select` or `Ident` ? If we are inside the namePos for select or within Ident pos then we should be good

Comment (harpocrates) [reply mtags/src/main/scala-2/scala/meta/internal/pc/PcDefinitionProvider.scala:206]: Sorry for the slow reply. Trying this out now, but have a couple questions:

  * Is `namePos` actually defined for 2.12.17?
  * On 2.13.6, even with `namePos`, doesn't that include space around the name (ex `tests.typedef.TypeDefinitionSuite.whitespace_2.13.16` would fail - `def x = 1 @@ + 2` places the `@@` cursor inside the `+` from the select node)

Comment (tgodzik) [reply mtags/src/main/scala-2/scala/meta/internal/pc/PcDefinitionProvider.scala:206]: Ach, I think we needed to rename it to `namePosition` it's an implicit class on DefTree.

It should not have the same issue as namePos, but not 100% sure

Comment (harpocrates) [reply mtags/src/main/scala-2/scala/meta/internal/pc/PcDefinitionProvider.scala]: Oh, that's very useful! Updated and removed all of the tokenization. I ended up needing to tweak a bit the case for imports - there we were producing our own synthetic `Ident` AST node, but it had the wrong span.

**Comment by harpocrates** [comment]
I'm not able to repro the CI failures locally. I also can't tell if they are flakes 🤔

**Comment by tgodzik** [comment]
> I'm not able to repro the CI failures locally. I also can't tell if they are flakes 🤔

I managed to reproduce after rebasing and I think the issue is in the incompatibility between two ways of find definition that we have. Not really related to this PR, but it would show up here because of the removed conditions

**Review by tgodzik — APPROVED** [review]
Looks good! Thanks for the fix! Let's hope the tests pass now (aside from the everpresent flaky ones)
